src/pyqasm/modules/base.py

```diff
@@ -761,3 +763,162 @@ def accept(self, visitor):
+            if isinstance(stmt, qasm3_ast.QuantumGate):
+                # Offset all qubit operands
+                for q in stmt.qubits:
+                    for group in q.indices:
+                        for ind in group:
+                            ind.value += offset  # type: ignore[attr-defined]
+                return
+
+            if isinstance(stmt, qasm3_ast.QuantumReset):
+                q = stmt.qubits
+                if isinstance(q, qasm3_ast.IndexedIdentifier):
+                    for group in q.indices:
+                        for ind in group:
+                            ind.value += offset  # type: ignore[attr-defined]
+                return
+
+            if isinstance(stmt, qasm3_ast.QuantumBarrier):
+                # Barrier can be represented with IndexedIdentifier or a string slice on Identifier
+                qubits = stmt.qubits
+                if len(qubits) == 0:
+                    return
+                first = qubits[0]
+                if isinstance(first, qasm3_ast.IndexedIdentifier):
+                    for group in first.indices:
+                        for ind in group:
+                            ind.value += offset  # type: ignore[attr-defined]
+                elif isinstance(first, qasm3_ast.Identifier):
+                    # Handle forms: __PYQASM_QUBITS__[:E], [S:], [S:E]
+                    name = first.name
+                    if name.startswith("__PYQASM_QUBITS__[") and name.endswith("]"):
+                        slice_str = name[len("__PYQASM_QUBITS__"):]
+                        # Parse slice forms [S:E], [:E], or [S:] and capture optional start/end integers
+                        m = re.match(r"\[(?:(\d+)?:(\d+)?)\]", slice_str)
+                        if m:
+                            start_s, end_s = m.group(1), m.group(2)
+                            if start_s is None and end_s is not None:
+                                # [:E]
+                                end_v = int(end_s) + offset
+                                first.name = f"__PYQASM_QUBITS__[:{end_v}]"
+                            elif start_s is not None and end_s is None:
+                                # [S:]
+                                start_v = int(start_s) + offset
+                                first.name = f"__PYQASM_QUBITS__[{start_v}:]"
+                            elif start_s is not None and end_s is not None:
+                                # [S:E]
+                                start_v = int(start_s) + offset
+                                end_v = int(end_s) + offset
+                                first.name = f"__PYQASM_QUBITS__[{start_v}:{end_v}]"
+                return
+
+        # Append statements with index offset, skipping its qubit declaration and include statements
+        for stmt in right_mod.unrolled_ast.statements:
+            if isinstance(stmt, (qasm3_ast.QubitDeclaration, qasm3_ast.Include)):
+                continue
+            stmt_copy = deepcopy(stmt)
```

Comment:
Don't think we'll need a copy of the statements again as we are already working on the copy of the module itself